README.md
# Flutter menu drawer 

[![pub package](https://img.shields.io/badge/pub-1.0.3-blue)](https://github.com/fahmyabada/menu_drawer)

A Flutter package with custom implementation of Menu Drawer 

<img src="https://github.com/fahmyabada/menu_drawer/raw/master/screenshot_1.png" alt="screenshot" width="250" height="481"/>


To start using this package, add `menu_drawer` dependency to your `pubspec.yaml`

```yaml
dependencies:
  menu_drawer: '<latest_release>'
```

# Features

- control header menu
- add many items (all control leading - all control name - press)
- add power by
- add versions

# how to use

```
@override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      drawer: Menu(
        item: [
          ListTile(
            title: const Text(
              'Home',
              style: TextStyle(
                color: Colors.grey,
                fontSize: 20,
              ),
            ),
            leading: const Icon(
              Icons.home,
              color: Colors.grey,
              size: 25,
            ),
            onTap: () {
              Navigator.pop(context);
            },
          ),
          Divider(
            height: 1,
            color: Colors.grey[400],
          ),
          const SizedBox(
            height: 10,
          ),
          ListTile(
            title: const Text(
              'Support',
              style: TextStyle(
                color: Colors.grey,
                fontSize: 20,
              ),
            ),
            leading: const Icon(
              Icons.support,
              color: Colors.grey,
              size: 25,
            ),
            onTap: () {
              Navigator.pop(context);
            },
          ),
          Divider(
            height: 1,
            color: Colors.grey[400],
          ),
          const SizedBox(
            height: 10,
          ),
          ListTile(
            title: const Text(
              'About',
              style: TextStyle(
                color: Colors.grey,
                fontSize: 20,
              ),
            ),
            leading: const Icon(
              Icons.abc_outlined,
              color: Colors.grey,
              size: 25,
            ),
            onTap: () {
              Navigator.pop(context);
            },
          ),
          Divider(
            height: 1,
            color: Colors.grey[400],
          ),
          const SizedBox(
            height: 10,
          ),
        ],
        titleDrawerHeader: "mohamed fahmy",
        width: 320.0,
        colorDrawerHeader: Colors.blue,
        heightDrawerHeader: 70.0,
        widthDrawerHeader: 70.0,
        urlImageDrawerHeader:
        "https://www.escapeauthority.com/wp-content/uploads/2116/11/No-image-found.jpg",
        spaceDrawerHeader: 15.0,
        colorTitleDrawerHeader: Colors.white,
        fontSizeTitleDrawerHeader: 20.0,
        subTitleDrawerHeader: "",
        colorSubTitleDrawerHeader: Colors.white,
        fontSizeSubTitleDrawerHeader: 20.0,
        namePoweredBy: 'PoweredBy',
        colorPoweredBy: Colors.blue,
        fontSizePoweredBy: 17.0,
        heightNameImagePoweredBy: 5,
        pressImagePoweredBy: () {
          print('pressImagePoweredBy');
        },
        urlImagePoweredBy: "assets/logoCompany.png",
        heightImagePoweredBy: 40.0,
        widthImagePoweredBy: 150.0,
        heightImageVersion: 10.0,
        nameVersion: 'Version',
        valueVersion: "1.0",
        colorVersion: Colors.blue,
        fontSizeVersion: 17.0,
      ),
    );
  }
```

License
----

BSD 3-clause License


CHANGELOG.md
## [1.0.0] - Release

* sample menu

## [1.0.1] - Release

* modify example 
* change project to private

## [1.0.2] - Release

## [1.0.3] - Release


example_menu_drawer/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:example_menu_drawer/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/menu.dart
library menu_drawer;

import 'package:flutter/material.dart';
import 'package:menu_drawer/image_tools.dart';

class Menu extends StatelessWidget {
  // with menu
  final double width;

  // all attribute DrawerHeader
  final Color? colorDrawerHeader;
  final double? heightDrawerHeader;
  final double? widthDrawerHeader;
  final String? urlImageDrawerHeader;

  // space between image and title in DrawerHeader
  final double? spaceDrawerHeader;
  final String? titleDrawerHeader;
  final Color? colorTitleDrawerHeader;
  final double? fontSizeTitleDrawerHeader;
  final String? subTitleDrawerHeader;
  final Color? colorSubTitleDrawerHeader;
  final double? fontSizeSubTitleDrawerHeader;

  // group of widget you want show ex: about page - privacyPolicy page ..
  final List<Widget> item;

  // all attribute PoweredBy
  final String? namePoweredBy;
  final Color? colorPoweredBy;
  final double? fontSizePoweredBy;

  // space between Name of PoweredBy and Image PoweredBy
  final double? heightNameImagePoweredBy;
  final GestureTapCallback? pressImagePoweredBy;
  final String? urlImagePoweredBy;
  final double? heightImagePoweredBy;
  final double? widthImagePoweredBy;

  // all attribute Version
  final double? heightImageVersion;
  final String? nameVersion;
  final String? valueVersion;
  final Color? colorVersion;
  final double? fontSizeVersion;

  const Menu({
    super.key,
    required this.titleDrawerHeader,
    required this.item,
    this.width = 320.0,
    this.colorDrawerHeader = Colors.blue,
    this.heightDrawerHeader = 70.0,
    this.widthDrawerHeader = 70.0,
    this.urlImageDrawerHeader,
    this.spaceDrawerHeader = 15.0,
    this.colorTitleDrawerHeader = Colors.white,
    this.fontSizeTitleDrawerHeader = 20.0,
    this.subTitleDrawerHeader = "",
    this.colorSubTitleDrawerHeader = Colors.white,
    this.fontSizeSubTitleDrawerHeader = 20.0,
    this.namePoweredBy = 'PoweredBy',
    this.colorPoweredBy = Colors.blue,
    this.fontSizePoweredBy = 17.0,
    this.heightNameImagePoweredBy = 5,
    this.pressImagePoweredBy,
    this.urlImagePoweredBy = "assets/logoCompany.png",
    this.heightImagePoweredBy = 40.0,
    this.widthImagePoweredBy = 150.0,
    this.heightImageVersion = 10.0,
    this.nameVersion = 'Version',
    this.valueVersion = "1.0",
    this.colorVersion = Colors.blue,
    this.fontSizeVersion = 17.0,
  });

  @override
  Widget build(BuildContext context) {
    return Drawer(
      width: width,
      child: CustomScrollView(
        slivers: [
          SliverToBoxAdapter(
            child: DrawerHeader(
              decoration: BoxDecoration(color: colorDrawerHeader),
              child: Row(
                children: [
                  SizedBox(
                    height: heightDrawerHeader,
                    width: widthDrawerHeader,
                    child: ClipOval(
                      child: ImageTools.image(
                        fit: BoxFit.fill,
                        url: urlImageDrawerHeader ?? "",
                      ),
                    ),
                  ),
                  SizedBox(
                    width: spaceDrawerHeader,
                  ),
                  Expanded(
                    flex: 4,
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: [
                        Text(
                          titleDrawerHeader ?? "",
                          maxLines: 1,
                          style: TextStyle(
                              overflow: TextOverflow.ellipsis,
                              color: colorTitleDrawerHeader,
                              fontSize: fontSizeTitleDrawerHeader,
                              fontWeight: FontWeight.bold),
                        ),
                        subTitleDrawerHeader!.isNotEmpty
                            ? Text(
                                subTitleDrawerHeader ?? "",
                                maxLines: 2,
                                textDirection: TextDirection.ltr,
                                style: TextStyle(
                                  overflow: TextOverflow.ellipsis,
                                  color: colorSubTitleDrawerHeader,
                                  fontSize: fontSizeSubTitleDrawerHeader,
                                ),
                              )
                            : Container(),
                      ],
                    ),
                  ),
                ],
              ),
            ),
          ),
          item.isNotEmpty
              ? SliverList.list(
                  children: item,
                )
              : SliverToBoxAdapter(
                  child: Container(),
                ),
          SliverFillRemaining(
            hasScrollBody: false,
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.center,
              children: [
                Text(
                  namePoweredBy ?? 'PoweredBy',
                  style: TextStyle(
                    color: colorPoweredBy,
                    fontSize: fontSizePoweredBy,
                  ),
                ),
                SizedBox(
                  height: heightNameImagePoweredBy,
                ),
                InkWell(
                  onTap: pressImagePoweredBy,
                  child: ImageTools.image(
                    fit: BoxFit.contain,
                    url: urlImagePoweredBy,
                    width: widthImagePoweredBy,
                    height: heightImagePoweredBy,
                  ),
                ),
                SizedBox(
                  height: heightImageVersion,
                ),
                Text(
                  '$nameVersion $valueVersion',
                  style: TextStyle(
                    color: colorVersion,
                    fontSize: fontSizeVersion,
                  ),
                ),
                const SizedBox(
                  height: 20.0,
                ),
              ],
            ),
          )
        ],
      ),
    );
  }
}


lib/image_tools.dart
import 'dart:io';

import 'package:cached_network_image/cached_network_image.dart';
import 'package:extended_image/extended_image.dart';
import 'package:flutter/material.dart';

// ignore: camel_case_types
enum kSize { small, medium, large }

class ImageTools {
  static String prestashopImage(String url, [kSize? size = kSize.medium]) {
    if (url.contains('?')) {
      switch (size) {
        case kSize.large:
          return url.replaceFirst('?', '/large_default?');
        case kSize.small:
          return url.replaceFirst('?', '/small_default?');
        default: // kSize.medium
          return url.replaceFirst('?', '/medium_default?');
      }
    }
    switch (size) {
      case kSize.large:
        return '$url/large_default';
      case kSize.small:
        return '$url/small_default';
      default: // kSize.medium
        return '$url/medium_default';
    }
  }

  static NetworkImage networkImage(String? url, [kSize size = kSize.medium]) {
    return NetworkImage(url!);
  }

  /// Smart image function to load image cache and check empty URL to return empty box
  /// Only apply for the product image resize with (small, medium, large)
  static Widget image({
    String? url,
    kSize? size,
    double? width,
    double? height,
    BoxFit? fit,
    String? tag,
    int? cacheHeight,
    int? cacheWidth,
    double offset = 0.0,
    bool isResize = false,
    bool? isVideo = false,
    bool hidePlaceHolder = false,
    bool forceWhiteBackground = false,
    String kImageProxy = '',
  }) {
    // if (height == null && width == null) {
    //   width = 20;
    // }
    var ratioImage = 1.4;

    if (url?.isEmpty ?? true) {
      // debugPrint('image*****************111 =');

      return ExtendedImage.network(
        "https://www.pphfoundation.ca/wp-content/uploads/2018/05/default-avatar.png",
        width: width,
        height: height,
        fit: fit,
        cache: true,
        enableLoadState: true,
        loadStateChanged: (ExtendedImageState state) {
          Widget? widget;
          switch (state.extendedImageLoadState) {
            case LoadState.loading:
              // showLoadingDialog();
              break;
            case LoadState.completed:
              widget = ExtendedRawImage(
                image: state.extendedImageInfo?.image,
                width: width,
                height: height,
                fit: fit,
              );
              break;
            case LoadState.failed:
              widget = Container(
                width: width,
                height: height ?? width! * ratioImage,
                color: const Color(0x0D000000),
              );
              break;
          }
          return widget;
        },
      );
    }
    else if (url!.startsWith("assets")) {
      // debugPrint('image*****************222 =');
      return Image.asset(
        url,
        width: width,
        height: height,
        fit: fit,
      );
    }
    else if (url.startsWith("http")) {
      // debugPrint('image*****************333 = $url');
      return ExtendedImage.network(
        url,
        width: width,
        height: height,
        fit: fit,
        cache: true,
        enableLoadState: true,
        cacheHeight: cacheHeight,
        cacheWidth: cacheWidth,
        loadStateChanged: (ExtendedImageState state) {
          Widget? widget;
          switch (state.extendedImageLoadState) {
            case LoadState.loading:
              // showLoadingDialog();
              break;
            case LoadState.completed:
              widget = ExtendedRawImage(
                image: state.extendedImageInfo?.image,
                width: width,
                height: height,
                fit: fit,
              );
              break;
            case LoadState.failed:
              widget = Container(
                width: width,
                height: height ?? width! * ratioImage,
                color: const Color(0x0D000000),
              );
              break;
          }
          return widget;
        },
      );
    } else {
      // debugPrint('image*****************444 =');
      return Image.file(
        File(url),
        height: height,
        fit: fit,
      );
    }
  }

  /// cache avatar for the chat
  static CachedNetworkImage getCachedAvatar(String avatarUrl) {
    return CachedNetworkImage(
      imageUrl: avatarUrl,
      imageBuilder: (context, imageProvider) => CircleAvatar(
        backgroundImage: imageProvider,
      ),
      placeholder: (context, url) => const CircularProgressIndicator(),
      errorWidget: (context, url, error) => const Icon(Icons.error),
    );
  }

  static BoxFit boxFit(String? fit) {
    switch (fit) {
      case 'contain':
        return BoxFit.contain;
      case 'fill':
        return BoxFit.fill;
      case 'fitHeight':
        return BoxFit.fitHeight;
      case 'fitWidth':
        return BoxFit.fitWidth;
      case 'scaleDown':
        return BoxFit.scaleDown;
      default:
        return BoxFit.cover;
    }
  }
}


example/linux/my_application.cc
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));

  // Use a header bar when running in GNOME as this is the common style used
  // by applications and is the setup most users will be using (e.g. Ubuntu
  // desktop).
  // If running on X and not using GNOME then just use a traditional title bar
  // in case the window manager does more exotic layout, e.g. tiling.
  // If running on Wayland assume the header bar will work (may need changing
  // if future cases occur).
  gboolean use_header_bar = TRUE;
#ifdef GDK_WINDOWING_X11
  GdkScreen* screen = gtk_window_get_screen(window);
  if (GDK_IS_X11_SCREEN(screen)) {
    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
      use_header_bar = FALSE;
    }
  }
#endif
  if (use_header_bar) {
    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
    gtk_widget_show(GTK_WIDGET(header_bar));
    gtk_header_bar_set_title(header_bar, "example_menu_drawer");
    gtk_header_bar_set_show_close_button(header_bar, TRUE);
    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  } else {
    gtk_window_set_title(window, "example_menu_drawer");
  }

  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_grab_focus(GTK_WIDGET(view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application, gchar*** arguments, int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
     g_warning("Failed to register: %s", error->message);
     *exit_status = 1;
     return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line = my_application_local_command_line;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID,
                                     "flags", G_APPLICATION_NON_UNIQUE,
                                     nullptr));
}


example_menu_drawer/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.Create(L"example_menu_drawer", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:menu_drawer/menu.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // TRY THIS: Try running your application with "flutter run". You'll see
        // the application has a blue toolbar. Then, without quitting the app,
        // try changing the seedColor in the colorScheme below to Colors.green
        // and then invoke "hot reload" (save your changes or press the "hot
        // reload" button in a Flutter-supported IDE, or press "r" if you used
        // the command line to start the app).
        //
        // Notice that the counter didn't reset back to zero; the application
        // state is not lost during the reload. To reset the state, use hot
        // restart instead.
        //
        // This works for code too, not just values: Most code changes can be
        // tested with just a hot reload.
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const ExampleApp(),
    );
  }
}

class ExampleApp extends StatelessWidget {
  const ExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      drawer: Menu(
        item: [
          ListTile(
            title: const Text(
              'Home',
              style: TextStyle(
                color: Colors.grey,
                fontSize: 20,
              ),
            ),
            leading: const Icon(
              Icons.home,
              color: Colors.grey,
              size: 25,
            ),
            onTap: () {
              Navigator.pop(context);
            },
          ),
          Divider(
            height: 1,
            color: Colors.grey[400],
          ),
          const SizedBox(
            height: 10,
          ),
          ListTile(
            title: const Text(
              'Support',
              style: TextStyle(
                color: Colors.grey,
                fontSize: 20,
              ),
            ),
            leading: const Icon(
              Icons.support,
              color: Colors.grey,
              size: 25,
            ),
            onTap: () {
              Navigator.pop(context);
            },
          ),
          Divider(
            height: 1,
            color: Colors.grey[400],
          ),
          const SizedBox(
            height: 10,
          ),
          ListTile(
            title: const Text(
              'About',
              style: TextStyle(
                color: Colors.grey,
                fontSize: 20,
              ),
            ),
            leading: const Icon(
              Icons.abc_outlined,
              color: Colors.grey,
              size: 25,
            ),
            onTap: () {
              Navigator.pop(context);
            },
          ),
          Divider(
            height: 1,
            color: Colors.grey[400],
          ),
          const SizedBox(
            height: 10,
          ),
        ],
        titleDrawerHeader: "mohamed fahmy",
        width: 320.0,
        colorDrawerHeader: Colors.blue,
        heightDrawerHeader: 70.0,
        widthDrawerHeader: 70.0,
        urlImageDrawerHeader:
        "https://www.escapeauthority.com/wp-content/uploads/2116/11/No-image-found.jpg",
        spaceDrawerHeader: 15.0,
        colorTitleDrawerHeader: Colors.white,
        fontSizeTitleDrawerHeader: 20.0,
        subTitleDrawerHeader: "",
        colorSubTitleDrawerHeader: Colors.white,
        fontSizeSubTitleDrawerHeader: 20.0,
        namePoweredBy: 'PoweredBy',
        colorPoweredBy: Colors.blue,
        fontSizePoweredBy: 17.0,
        heightNameImagePoweredBy: 5,
        pressImagePoweredBy: () {
          print('pressImagePoweredBy');
        },
        urlImagePoweredBy: "assets/logoCompany.png",
        heightImagePoweredBy: 40.0,
        widthImagePoweredBy: 150.0,
        heightImageVersion: 10.0,
        nameVersion: 'Version',
        valueVersion: "1.0",
        colorVersion: Colors.blue,
        fontSizeVersion: 17.0,
      ),
    );
  }
}
